### PR TITLE
add audio models: audio_convnet and audio_conv_rnn

### PR DESCRIPTION
**Pull request by keunwoochoi**
.h5 files are now at https://github.com/keunwoochoi/music-auto_tagging-keras/tree/master/data.

### CODE REVIEW

**Comment by fchollet** [comment]
Please add the code for preprocessing new audio tracks, as well as code for converting the network's predictions back to readable tags (e.g. same as https://github.com/fchollet/keras/blob/master/keras/applications/imagenet_utils.py#L11 and https://github.com/fchollet/keras/blob/master/keras/applications/imagenet_utils.py#L31).

Assuming librosa is available is fine. If not available, a clear error message should be displayed to tell the user to go install it. The helper functions should be available as `audio_convnet.preprocess_input` and `audio_convnet.decode_predictions` (resp. `audio_conv_rnn`) to be consistent with other applications.

Additionally, please add documentation in `docs/templates/applications.md` including a code example for tagging a new MP3 file.

**Comment by keunwoochoi** [comment]
1. `preprocess_input()` would be the same for both of them. You want it to be not in separate file (`audio_utils.py`) but copy&pasted in both `audio_convnet.py` and `audio_conv_rnn.py`?

2. I'll make `decode_predictions()` to return a list of tuples, `(tag(string), prob(float))` e.g. `[('folk', 0.222), ('pop', 0.166), ('jazz', 0.160), ('female vocalists', 0.092), ('acoustic', 0.075)]`. Is it good?

**Comment by keunwoochoi** [comment]
+ I'd like the API to be `preprocess_input(audio_path)` rather than `preprocess_input(x)` (`x`:1-d numpy array for raw-audio) as reading audio is much easier with librosa and anyway librosa is required to compute spectrograms. This makes the API bit different to previous imagenet utils though.

**Comment by fchollet** [comment]
> I'd like the API to be preprocess_input(audio_path) rather than preprocess_input(x) (x:1-d numpy array for raw-audio) as reading audio is much easier with librosa 

That's fine.

> I'll make decode_predictions() to return a list of tuples,

That's fine, but note that it should be able to process a batch of predictions, so the output would be a list of lists.

> preprocess_input() would be the same for both of them. You want it to be not in separate file (audio_utils.py) but copy&pasted in both audio_convnet.py and audio_conv_rnn.py?

You can do it the same way it is done for vision models: the code is in a separate file, but each model module imports util functions from there so they are available as module attrs: https://github.com/fchollet/keras/blob/master/keras/applications/resnet50.py#L23

**Comment by fchollet** [comment]
Please fix trailing whitespace PEP8 errors.

File: docs/templates/applications.md [hunk 1]
```diff
@@ -151,6 +153,31 @@ input_tensor = Input(shape=(224, 224, 3))  # this assumes K.image_dim_ordering()
...
 
+
+### Tag music with AudioConvnet
+
+```python
+
+from keras.applications.audio_convnet import AudioConvnet
+from keras.applications.audio_convnet import load_preprocess_input, decode_predictions
+
+# this could also be the output a different Keras model or layer
```
Comment (fchollet) [inline docs/templates/applications.md]: ?

File: keras/applications/audio_conv_rnn.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        freq_axis = 1
+        time_axis = 2
+
+    # Input block
+    x = ZeroPadding2D(padding=(0, 37))(melgram_input)
+    x = BatchNormalization(axis=time_axis, name='bn_0_freq')(x)
+
+    # Conv block 1
+    x = Convolution2D(64, 3, 3, border_mode='same', name='conv1')(x)
+    x = BatchNormalization(axis=channel_axis, mode=2, name='bn1')(x)
```
Comment (fchollet) [inline keras/applications/audio_conv_rnn.py]: Only noticing this now, but why use BN in mode 2? This should be mode 0...

File: keras/applications/audio_convnet.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        channel_axis = 3
+        freq_axis = 1
+        time_axis = 2
+
+    # Input block
+    x = BatchNormalization(axis=time_axis, name='bn_0_freq')(melgram_input)
+
+    # Conv block 1
+    x = Convolution2D(32, 3, 3, border_mode='same', name='conv1')(x)
+    x = BatchNormalization(axis=channel_axis, mode=2, name='bn1')(x)
```
Comment (fchollet) [inline keras/applications/audio_convnet.py]: Same here --surely mode 0 would be preferable?

File: keras/applications/audio_conv_rnn.py (hunk 2 above)
Comment (keunwoochoi) [inline keras/applications/audio_conv_rnn.py]: Ah, it's been quite a long while and now I don't exactly remember why. And yes, as a result, the prediction is not very stable. I'll check it out with `mode=0` and PR again (+update weights if required)

Comment (keunwoochoi) [inline keras/applications/audio_conv_rnn.py]: Oh, a couple of months ago there seems a bug on `mode=0` -- at least I thought like that. 
Now I am training again with `mode=0` and it seems fine.

**Comment by carlthome** [comment]
`audio_convnet` and `audio_conv_rnn` are two very broad names and it's not clear what these two models map _from -> to_. Surely there must be a more concrete name that would explain this better?

**Comment by fchollet** [comment]
I will merge the PR when the BN mode is fixed (weights will need to be updated --in theory that doesn't require re-training, but the process of adding BN weights without retraining would be a bit of work too anyway).

Agreed with @carlthome that the names might benefit from being more specific (names are important since they will stick forever). I don't have any specific suggestion though. Do as you see fit...

**Comment by keunwoochoi** [comment]
Great. The new training is almost done now, will PR in 1-2 days. 
@carlthome They get mel-spectrogram and predict music tags. It would be too verbose to put melgram into name, but probably something like `music_cnn_tagger` and `music_rnn_tagger`? so that it specifies `music` rather than `audio` as well as the task (tagger).

**Comment by carlthome** [comment]
@keunwoochoi, yes, those names are a little better. I'd go for `_MODEL__music_autotagger` where _MODEL_ is your particular LSTM/CNN/RNN/whatever. There's a precedence for "autotagging". @douglaseck (lead on TensorFlow Magenta) used that terminology in [a paper of his](http://www.iro.umontreal.ca/~eckdoug/papers/2007_ismir.pdf) for example.

**Comment by keunwoochoi** [comment]
Hm, I'll think about it. 
* I put music_ as a comnon prefix as this is the most different aspect of then from other models.
* I removed 'auto' since it sounds redundant considering the context, I used auto-tagging in the title of my paper though. For me it is like putting 'Machine Learning' in the title in 2016 :)
* So probably they will be named as music_tagger_cnn(crnn) or music_cnn_tagger.

**Comment by fchollet** [comment]
> * So probably they will be named as music_tagger_cnn(crnn) or music_cnn_tagger.

Sounds good to me. Note that it should be formatted as MusicTaggerCNN / MusicTaggerCRNN (for consistency with the image models, and because they returns a class instance.

**Comment by keunwoochoi** [comment]
Okay.
A question: Is there `image_dim_ordering()` converter?

**Comment by keunwoochoi** [comment]
Can @fchollet or anyone test one thing? 
https://github.com/keunwoochoi/music-auto_tagging-keras/tree/mode0 **(Branch: mode0)**
This will be new PR - new model name, trained with BN `mode=0` and corresponding example (`example_new.py`).
On my macbook (keras 1.0.8, theano, MacOS) the prediction is float but either 0.0 or 1.0, which is weird. On my server (keras 1.0.6, theano, linux) the prediction is fine - as expected. I can't figure out why and if it's my problem or keras problem.

**Comment by fchollet** [comment]
@keunwoochoi running `example.py` in the repo you linked, branch `mode0` (after modifying it a bit to make it work with the new models), I get all-zero predictions. Theano, MacBook.

**Comment by keunwoochoi** [comment]
@fchollet Okay, I tested it more. First let me specify in detail. (I guess you actually ran `example_new.py`, right?)
Only Theano was used.

`example.py` : model with `mode=2` and trained in linux, keras 1.0.6.
`example_new.py`: model with `model=0` and trained in linux, keras, 1.0.6

* in mac, keras 1.0.6: Both of them works fine (shows valid results)
* in mac, keras 1.0.8: 
  - `example.py`: works well. However, the prediction is slightly different from 1.0.6 (e.g., 0.226 vs 0.222)
  - `example_new.py`: predicting only either 0.0 or 1.0

Do you have any idea?

**Comment by fchollet** [comment]
> Do you have any idea?

The short answer: the predictions at the `flatten` step are in the range [-1, 500], roughly. Typically low 100s. The last `Dense` layer maps them to a space where all predictions are either negative ~5000-10000 or positive ~5000-1000 (rare case, generally all negative). So the sigmoid maps them to 0 or 1.

There were some changes to the BN layers between 1.0.6 and 1.0.8. Maybe it's related. The changes were all backwards compatible though.

**Comment by keunwoochoi** [comment]
> The short answer:
And the results say that happens only in keras 1.0.8, which means they are not backward-compatible in this case? 

I can just update the keras on my server to 1.0.8 and run it again, but probably you also wanna fix it if there's a problem. Let me know if you anyway want me to run it again.

**Comment by fchollet** [comment]
First we would need to know if and when the problem arises... 

This is probably the commit we need to look into: https://github.com/fchollet/keras/commit/97d2a73dd36e90a3502c575a8ea93125e894c84a

The current version is most definitely correct; the previous version wasn't.

**Comment by keunwoochoi** [comment]
> This is probably

It seems to me as well. Probably the `std`(in weights learned in 1.0.6) should be changed to `std ** 2` to be used in 1.0.8? I'll test it.
I will train it again anyway.

**Comment by keunwoochoi** [comment]
PS. It should be the case then all the other weights files should be also updated if they were using BN on Theano, right? Hm..

**Comment by keunwoochoi** [comment]
Not sure if it is related, but CRNN works fine with keras 1.1.0 (+with the weights that are trained by keras 1.0.6). I also added `no_top` option on both of them.
I will PR again only with CRNN first, while looking into CNN model more (or just re-train them).

**Comment by keunwoochoi** [comment]
related: https://github.com/keunwoochoi/music-auto_tagging-keras/blob/mode0/music_tagger_crnn.py, 
https://github.com/keunwoochoi/music-auto_tagging-keras/blob/mode0/example_feat_extract.py

**Comment by keunwoochoi** [comment]
The current version (6cadc7b) : `MusicTaggerCRNN` only, application.md is updated, `include_top` argument is added, no issue on keras version with BN layers.

File: keras/applications/audio_conv_utils.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+def librosa_exists():
+    try:
+        __import__('librosa')
+    except ImportError:
+        return False
+    else:
+        return True
+
+
+def load_preprocess_input(audio_path, dim_ordering='default'):
```
Comment (fchollet) [inline keras/applications/audio_conv_utils.py]: I would argue that it is fine to name this `preprocess_input` alone.

File: keras/applications/audio_conv_utils.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    if dim_ordering == 'default':
+        dim_ordering = K.image_dim_ordering()
+    assert dim_ordering in {'tf', 'th'}
+
+    # mel-spectrogram parameters
+    SR = 12000
+    N_FFT = 512
+    N_MELS = 96
+    HOP_LEN = 256
+    DURA = 29.12
```
Comment (fchollet) [inline keras/applications/audio_conv_utils.py]: Do you think it would make sense to make these into keyword arguments of the function? Or do they have to be constants?

File: keras/applications/audio_conv_utils.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    src, sr = librosa.load(audio_path, sr=SR)
+    n_sample = src.shape[0]
+    n_sample_wanted = int(DURA*SR)
```
Comment (fchollet) [inline keras/applications/audio_conv_utils.py]: PEP8

File: keras/applications/audio_conv_utils.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    # trim the signal at the center
+    if n_sample < n_sample_wanted:  # if too short
+        src = np.hstack((src, np.zeros((int(DURA*SR) - n_sample,))))
```
Comment (fchollet) [inline keras/applications/audio_conv_utils.py]: PEP8

File: keras/applications/audio_conv_utils.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    elif n_sample > n_sample_wanted:  # if too long
+        src = src[(n_sample-n_sample_wanted)/2:(n_sample+n_sample_wanted)/2]
```
Comment (fchollet) [inline keras/applications/audio_conv_utils.py]: PEP8

File: keras/applications/audio_conv_utils.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    logam = librosa.logamplitude
+    melgram = librosa.feature.melspectrogram
+    x = logam(melgram(y=src, sr=SR, hop_length=HOP_LEN,
+                      n_fft=N_FFT, n_mels=N_MELS)**2,
```
Comment (fchollet) [inline keras/applications/audio_conv_utils.py]: PEP8

File: keras/applications/audio_conv_utils.py (hunk 4 above)
Comment (keunwoochoi) [inline keras/applications/audio_conv_utils.py]: Okay, I don't have a strong opinion and it sounds good.

File: keras/applications/audio_conv_utils.py (hunk 5 above)
Comment (keunwoochoi) [inline keras/applications/audio_conv_utils.py]: They have to be constants.

File: keras/applications/audio_conv_utils.py (hunk 6 above)
Comment (keunwoochoi) [inline keras/applications/audio_conv_utils.py]: `pep8` doesn't show me any message on this. I presume it should be `(DURA * SR)`?

File: keras/applications/audio_conv_utils.py (hunk 7 above)
Comment (keunwoochoi) [inline keras/applications/audio_conv_utils.py]: probably same here.

File: keras/applications/audio_conv_utils.py (hunk 8 above)
Comment (keunwoochoi) [inline keras/applications/audio_conv_utils.py]: If it's the same, should it be
```
src = src[(n_sample - n_sample_wanted)/2:
                  (n_sample + n_sample_wanted)/2]
```
?

File: keras/applications/audio_conv_utils.py (hunk 9 above)
Comment (keunwoochoi) [inline keras/applications/audio_conv_utils.py]: Could you elaborate here?

Comment (fchollet) [inline keras/applications/audio_conv_utils.py]: Operators (such as `**`) should have spaces around them.

File: keras/applications/audio_conv_utils.py (hunk 8 above)
Comment (fchollet) [inline keras/applications/audio_conv_utils.py]: Or even better:

```python
src = src[(n_sample - n_sample_wanted) / 2:
               (n_sample + n_sample_wanted) / 2]

```

**Comment by fchollet** [comment]
LGTM, thanks!

**Comment by fchollet** [comment]
@keunwoochoi I made a few fixes I didn't catch at review time: https://github.com/fchollet/keras/commit/ed131973ef5748915ab3a03c9f60c9051d34881c

It seems to me like it shouldn't be an issue to convert the checkpoint to work with `image_dim_ordering = 'tf'`. It would just require transposing every conv2d kernel. What do you think?

**Comment by keunwoochoi** [comment]
@fchollet I was not sure how can I change the stored weights if `image_dim_ordering='tf'`. And yes, it is only about transposing every conv2d kernel. It would be much better if it works for both.

**Comment by fchollet** [comment]
@keunwoochoi I added support for both dim orderings.
